controller/volume_controller.go

```diff
@@ -1277,401 +1591,238 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
-			v.Status.Robustness = longhorn.VolumeRobustnessUnknown
-			// The volume was faulty and there are usable replicas.
-			// Therefore, we set RemountRequestedAt so that KubernetesPodController restarts the workload pod
-			v.Status.RemountRequestedAt = vc.nowHandler()
-			msg := fmt.Sprintf("Volume %v requested remount at %v", v.Name, v.Status.RemountRequestedAt)
-			vc.eventRecorder.Eventf(v, v1.EventTypeNormal, constant.EventReasonRemount, msg)
+		rs[r.Name] = r
+	}
+
+	replicaAddressMap := map[string]string{}
+	for _, r := range rs {
+		// Ignore unscheduled replicas
+		if r.Spec.NodeID == "" {
+			continue
+		}
+		if r.Spec.EngineImage != v.Status.CurrentImage {
+			continue
+		}
+		if r.Spec.EngineName != e.Name {
+			continue
+		}
+		if r.Spec.FailedAt != "" {
+			continue
+		}
+		if r.Status.CurrentState == longhorn.InstanceStateError {
+			continue
 		}
+		// wait for all potentially healthy replicas become running
+		if r.Status.CurrentState != longhorn.InstanceStateRunning {
+			return nil
+		}
+		if r.Status.IP == "" {
+			log.WithField("replica", r.Name).Error("BUG: replica is running but IP is empty")
+			continue
+		}
+		if r.Status.StorageIP == "" {
+			log.WithField("replica", r.Name).Debug("Replica is running but storage IP is empty, need to wait for update")
+			continue
+		}
+		if r.Status.Port == 0 {
+			log.WithField("replica", r.Name).Error("BUG: replica is running but Port is empty")
+			continue
+		}
+		replicaAddressMap[r.Name] = imutil.GetURL(r.Status.StorageIP, r.Status.Port)
+	}
+	if len(replicaAddressMap) == 0 {
+		return fmt.Errorf("no healthy or scheduled replica for starting")
+	}
 
-		// reattach volume if detached unexpected and there are still healthy replicas
-		if e.Status.CurrentState == longhorn.InstanceStateError && v.Status.CurrentNodeID != "" {
-			v.Status.PendingNodeID = v.Status.CurrentNodeID
-			log.Warn("Engine of volume dead unexpectedly, reattach the volume")
-			msg := fmt.Sprintf("Engine of volume %v dead unexpectedly, reattach the volume", v.Name)
-			vc.eventRecorder.Event(v, v1.EventTypeWarning, constant.EventReasonDetachedUnexpectly, msg)
-			e.Spec.LogRequested = true
-			for _, r := range rs {
-				if r.Status.CurrentState == longhorn.InstanceStateRunning {
-					r.Spec.LogRequested = true
-					rs[r.Name] = r
+	if e.Spec.NodeID != "" && e.Spec.NodeID != v.Spec.NodeID {
+		return fmt.Errorf("engine is on node %v vs volume on %v, must detach first",
+			e.Spec.NodeID, v.Status.CurrentNodeID)
+	}
+	e.Spec.NodeID = v.Spec.NodeID
+	e.Spec.ReplicaAddressMap = replicaAddressMap
+	e.Spec.DesireState = longhorn.InstanceStateRunning
+	// The volume may be activated
+	e.Spec.DisableFrontend = v.Status.FrontendDisabled
+	e.Spec.Frontend = v.Spec.Frontend
+
+	return nil
+}
+
+func (vc *VolumeController) verifyVolumeDependentResourcesOpened(e *longhorn.Engine, rs map[string]*longhorn.Replica) bool {
+	// At least 1 replica should be running
```

Comment:
Rename `areVolumeDependentResourcesOpened`

Reply:
Thank you. Fixed